### PR TITLE
ULTIMA8: Merge RenderSurface and SoftRenderSurface

### PR DESCRIPTION
**Pull request by OMGPizzaGuy**
Hi all! I'm doing these changes through a pull request to get feedback on them.
My aim is to simplify custom rendering in the engine and possibly eliminate methods that could be handled by ManagedSurface.
Should I flatten the render_surface.inl into the paint shape methods?
ShapeFrame owns a height, width, and pixels: should these be put in a Graphics::Surface before the custom blit or stored as a  surface in ShapeFrame directly?
Should I consider forcing the bpp on RenderSurface rather than letting it be configured by Ultima8Engine::GraphicSysInit?

Any other suggestions?

### CODE REVIEW

**Comment by sev-** [comment]
> ShapeFrame owns a height, width, and pixels: should these be put in a Graphics::Surface before the custom blit or stored as a surface in ShapeFrame directly?
I'd store Surface internally. 

Also, these `BlendedLogic` functions are suspiciously similar to the newly introduced `ManagedSurface::blendBlitTo()`, the former TransparentSurface code.

> Should I consider forcing the bpp on RenderSurface rather than letting it be configured by Ultima8Engine::GraphicSysInit?

Yes, this is always the best, as it simplifies the code significantly. Especially in your case, when there is only one game/family of games based on the engine, e.g. you know that some alien format will not arrive anyway.

**Comment by OMGPizzaGuy** [comment]
> > ShapeFrame owns a height, width, and pixels: should these be put in a Graphics::Surface before the custom blit or stored as a surface in ShapeFrame directly?
>  I'd store Surface internally.

Sounds good I'll rebase this branch and start down that path.

> Also, these `BlendedLogic` functions are suspiciously similar to the newly introduced `ManagedSurface::blendBlitTo()`, the former TransparentSurface code.

Indeed! Watching work on that inspired me to cycle back around to this as I wasn't sure how to tackle this previously.
 
> > Should I consider forcing the bpp on RenderSurface rather than letting it be configured by Ultima8Engine::GraphicSysInit?
> 
> Yes, this is always the best, as it simplifies the code significantly. Especially in your case, when there is only one game/family of games based on the engine, e.g. you know that some alien format will not arrive anyway.

Alright, I think I embrace the BlendBlit PixelFormat for this engine so I might be able to use the new code there.

Thanks!

**Comment by ccawley2011** [comment]
> > Should I consider forcing the bpp on RenderSurface rather than letting it be configured by Ultima8Engine::GraphicSysInit?
> 
> Yes, this is always the best, as it simplifies the code significantly. Especially in your case, when there is only one game/family of games based on the engine, e.g. you know that some alien format will not arrive anyway.

I should point out that there's usually a trade-off between simplicity and compatibility here since not all backends support the pixel format required by the common blending routines, and some of the ones that do perform an additional conversion pass in software that may introduce additional overhead on lower end machines.

I'd be interested to know what the Ultima8 engine is using alpha blending for - I wouldn't have expected a game from 1994-1996 to need it.

**Comment by OMGPizzaGuy** [comment]
> I should point out that there's usually a trade-off between simplicity and compatibility here since not all backends support the pixel format required by the common blending routines, and some of the ones that do perform an additional conversion pass in software that may introduce additional overhead on lower end machines.
>

Understood. I'm curious to know how well this engine performs on lower end as well, but have yet to check. There's a config key "bpp" currently to set 16 or 32 to see how much that pass would effect those machines.
 
> I'd be interested to know what the Ultima8 engine is using alpha blending for - I wouldn't have expected a game from 1994-1996 to need it.

Dragging items did a mask and blend in the original game to produce a light grey outline of the item. Street lights on ground , smoke effects from chimneys, and black potions for invisibility have similar blends  

Pentagram enhanced these as a full blending, which I think is a bit nicer. Also, there's a few extras for "Show Highlight Items" and "Show Touching Items" that blend colors - although those could be potentially handled by a modified pallete for those items.

**Comment by ccawley2011** [comment]
> Understood. I'm curious to know how well this engine performs on lower end as well, but have yet to check. There's a config key "bpp" currently to set 16 or 32 to see how much that pass would effect those machines.

What I'd suggest if possible is to select a format based on what the backend reports as supported. If the engine specifically requires RGB565 or RGBA8888, you can pass an array with those formats as the third parameter to `initGraphics`, while if it'll work with any 16bpp or 32bpp format, you can pass a null pointer instead and you'll get whatever true colour format the backend reports as being available and fast (with palette mode as a last resort).

> Dragging items did a mask and blend in the original game to produce a light grey outline of the item. Street lights on ground , smoke effects from chimneys, and black potions for invisibility have similar blends
> 
> Pentagram enhanced these as a full blending, which I think is a bit nicer. Also, there's a few extras for "Show Highlight Items" and "Show Touching Items" that blend colors - although those could be potentially handled by a modified pallete for those items.

That sounds like a neat enhancement. Does support for the original mask/blend effects exist in the current code, or is it just the enhanced version?

File: engines/ultima/ultima8/graphics/render_surface.cpp [hunk 1]
```diff
@@ -324,10 +292,131 @@ bool RenderSurface::IsFlipped() const {
...
+}
+
+} // End of anonymous namespace
+
+//
+// RenderSurface::FillAlpha(uint8 alpha, Rect r)
+//
+// Desc: Fill alpha channel
+//
+void RenderSurface::FillAlpha(uint8 alpha, const Rect &r) {
```
Comment (ccawley2011) [inline engines/ultima/ultima8/graphics/render_surface.cpp:355]: Am I correct in thinking that this is essentially the same as `Graphics::Surface::setAlpha()` except with an additional Rect parameter?

Comment (OMGPizzaGuy) [reply engines/ultima/ultima8/graphics/render_surface.cpp:355]: I believe you are correct. It was used in Pentagram for a two pass rendering of TextWidget for the normal scaled paint and then the composited unscaled paint to render a higher res font.

I've been maintaining this method and testing with CHECK_ALPHA_FILLS, although it's currently not used as the composited painting was abandoned when ported to ScummVM.

**Comment by OMGPizzaGuy** [comment]
> What I'd suggest if possible is to select a format based on what the backend reports as supported. If the engine specifically requires RGB565 or RGBA8888, you can pass an array with those formats as the third parameter to `initGraphics`, while if it'll work with any 16bpp or 32bpp format, you can pass a null pointer instead and you'll get whatever true colour format the backend reports as being available and fast (with palette mode as a last resort).

I'll keep this in mind.
 
> That sounds like a neat enhancement. Does support for the original mask/blend effects exist in the current code, or is it just the enhanced version?

Nope, never implemented an original version option.

**Comment by OMGPizzaGuy** [comment]
I think additional changes I'd like to make are starting to fall out of scope for this pull request. I'll hold off on those for the moment. Any additional review is appreciated.

Currently the benchmarks show a small improvement on my machine:
RenderSurface::benchmark 1 4 1000000
branch: master - 6c7467831d02cbb8cb40700f450c90de12ac91ef
[2023-08-28 18:15:28] Paint: 4444
[2023-08-28 18:15:32] PaintNoClip: 3737
[2023-08-28 18:15:37] PaintTranslucent: 4884
[2023-08-28 18:15:42] PaintMirrored: 4702
[2023-08-28 18:16:58] PaintInvisible: 76257
[2023-08-28 18:17:47] PaintHighlight: 49131
[2023-08-28 18:19:02] PaintHighlightInvis: 75299

PR #5288
[2023-09-02 11:46:32] Paint: 3843
[2023-09-02 11:46:32] PaintNoClip: N/A
[2023-09-02 11:46:36] PaintTranslucent: 4384
[2023-09-02 11:46:41] PaintMirrored: 4113
[2023-09-02 11:47:57] PaintInvisible: 76145
[2023-09-02 11:48:44] PaintHighlight: 47362
[2023-09-02 11:49:59] PaintHighlightInvis: 75079

File: engines/ultima/ultima8/graphics/render_surface.cpp [hunk 2]
```diff
@@ -116,16 +106,8 @@ bool RenderSurface::EndPainting() {
...
-			_pixels = _pixels00 = 0;
-
-			// Render the screen if this is it (slight hack..)
-			Graphics::Screen *screen = dynamic_cast<Graphics::Screen *>(_surface);
-			if (screen)
-				screen->update();
-		}
-		// else, nothing to unlock.
+		// Clear pointers
+		_pixels = 0;
```
Comment (mduggan) [inline engines/ultima/ultima8/graphics/render_surface.cpp:110]: minor nit, I think `nullptr` is slightly nicer than `0` here even though they have the same effect.

**Comment by mduggan** [comment]
Thanks for doing the cleanup! overall it LGTM - there are further improvements to be made but this should make them a lot easier!

File: engines/ultima/ultima8/graphics/render_surface.cpp (hunk 2 above)
Comment (OMGPizzaGuy) [reply engines/ultima/ultima8/graphics/render_surface.cpp]: Sure. Just amended the last commit to include that.

**Comment by OMGPizzaGuy** [comment]
I'll go ahead and complete this PR as I have a number of follow-up changes ready at this point.

**Pull request merged**